actors/builtin/power/testing.go

```diff
@@ -0,0 +1,232 @@
+		}
+
+		acc.Require(abi.ChainEpoch(epoch) >= st.FirstCronEpoch, "cron event at epoch %d before FirstCronEpoch %d",
+			epoch, st.FirstCronEpoch)
+
+		var event CronEvent
+		return arr.ForEach(&event, func(i int64) error {
+			byAddress[event.MinerAddr] = append(byAddress[event.MinerAddr], MinerCronEvent{
+				Epoch:   abi.ChainEpoch(epoch),
+				Payload: event.CallbackPayload,
+			})
+
+			return nil
+		})
+	})
+	acc.Require(err == nil, "error attempting to read through power actor cron tasks: %v", err)
+
+	return byAddress, nil
+}
+
+func CheckClaimInvariants(st *State, store adt.Store, acc *builtin.MessageAccumulator) (ClaimsByAddress, error) {
+	claims, err := adt.AsMap(store, st.Claims)
+	if err != nil {
+		return nil, err
+	}
+
+	committedRawPower := abi.NewStoragePower(0)
+	committedQAPower := abi.NewStoragePower(0)
+	rawPower := abi.NewStoragePower(0)
+	qaPower := abi.NewStoragePower(0)
+	claimsWithSufficientPowerCount := int64(0)
+	byAddress := make(ClaimsByAddress)
+	var claim Claim
+	err = claims.ForEach(&claim, func(key string) error {
+		addr, err := address.NewFromBytes([]byte(key))
+		if err != nil {
+			return err
+		}
+		byAddress[addr] = claim
+		committedRawPower = big.Add(committedRawPower, claim.RawBytePower)
+		committedQAPower = big.Add(committedQAPower, claim.QualityAdjPower)
+
+		minPower, err := builtin.ConsensusMinerMinPower(claim.SealProofType)
+		acc.Require(err == nil, "could not get consensus miner min power for miner %v: %v", addr, err)
+		if err != nil {
+			return nil // noted above
+		}
+
+		if claim.RawBytePower.GreaterThanEqual(minPower) {
+			claimsWithSufficientPowerCount += 1
+			rawPower = big.Add(rawPower, claim.RawBytePower)
+			qaPower = big.Add(qaPower, claim.QualityAdjPower)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	acc.Require(committedRawPower.Equals(st.TotalBytesCommitted),
+		"sum of raw power in claims %v does not match recorded bytes committed %v",
+		committedRawPower, st.TotalBytesCommitted)
+	acc.Require(committedQAPower.Equals(st.TotalQABytesCommitted),
+		"sum of qa power in claims %v does not match recorded qa power committed %v",
+		committedQAPower, st.TotalQABytesCommitted)
+
+	acc.Require(claimsWithSufficientPowerCount == st.MinerAboveMinPowerCount,
+		"claims with sufficient power %d does not match MinerAboveMinPowerCount %d",
+		claimsWithSufficientPowerCount, st.MinerAboveMinPowerCount)
+
+	acc.Require(st.TotalRawBytePower.Equals(rawPower),
+		"recorded raw power %v does not match raw power in claims %v", st.TotalRawBytePower, rawPower)
+	acc.Require(st.TotalQualityAdjPower.Equals(qaPower),
+		"recorded qa power %v does not match qa power in claims %v", st.TotalQualityAdjPower, qaPower)
+
+	return byAddress, nil
+}
+
+func CheckProofValidationInvariants(st *State, store adt.Store, claims ClaimsByAddress, acc *builtin.MessageAccumulator) (ProofsByAddress, error) {
+	if st.ProofValidationBatch == nil {
+		return nil, nil
+	}
```

Comment:
It is in fact an invariant that this queue is nil at the end of every epoch, because it is emptied by cron.

Reply:
Agreed, but the tests will sometimes run this before cron, so it might catch something.